sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -777,36 +772,84 @@ class Analyzer(
     }
 
     def apply(plan: LogicalPlan): LogicalPlan = ResolveTables(plan).resolveOperatorsUp {
-      case i @ InsertIntoStatement(u @ UnresolvedRelation(AsTableIdentifier(ident)), _, child, _, _)
-          if child.resolved =>
-        EliminateSubqueryAliases(lookupTableFromCatalog(ident, u)) match {
+      case i @ InsertIntoStatement(
+          u @ UnresolvedRelation(CatalogObjectIdentifier(catalog, ident)), _, _, _, _)
+            if i.query.resolved && CatalogV2Util.isSessionCatalog(catalog) =>
+        val relation = ResolveTempViews(u) match {
```

Comment:
why do we need this? temp views should always be resolve first. If we reach here, it's not a temp view.

Reply:
`u` is inside `InsertIntoStatement` (and not its child), so it is not resolved when we reach here.

Reply:
good catch! Can we resolve temp views inside `InsertIntoStatement` in `ResolveTempViews` as well?

Reply:
If we resolve temp views for `InsertIntoStatment.table` in `ResolveTempViews`, we need additional rule here to match `SubqueryAlias`. Is that what you were suggesting?

Reply:
We have `EliminateSubqueryAliases` here, so `SubqueryAlias` should be fine?

Reply:
Can you write to a temp view? Why is it valid for a temp view to be here?

Reply:
https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala#L87 is one example, where temp view resolution is required. Maybe the confusion is that `InsertIntoStatement` is used for `INSERT OVERWRITE`, `INSERT INTO`, etc.?

>  It seems to me that it should insert into the table default.t1 because it doesn't make sense to insert into the temp view.

I think it should still resolve to temp view (for consistent lookup behavior), but fails during analysis check, which is the current behavior.

Reply:
Okay, how does the analysis check catch the problem? Are we confident that always works?

Reply:
We have the following checks for `InsertIntoStatement`: https://github.com/apache/spark/blob/322ec0ba9ba75708cfe679368a43655de7b0e4f9/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala#L494

Reply:
Thanks! I think I would rather implement the check a different way:
1. Don't resolve temp tables in ResolveRelations
2. In CheckAnalysis (in catalyst), check for InsertInto(Unresolved(...)) and if the unresolved relation is a temp table state that it is a temp table and can't be resolved.

It would be good to know what other databases do in this case because my suggestion to not resolve the identifier as a temp table would allow matching a table here if there is one that conflicts. Probably good to consider this case in the broader clean up of temp table handling.

Reply:
> It would be good to know what other databases do in this case because my suggestion to not resolve the identifier as a temp table would allow matching a table here if there is one that conflicts. 

In Postgres, it is resolved to temp view:

```
postgres=# create schema s1;
CREATE SCHEMA
postgres=# SET search_path TO s1;
SET
postgres=# create table s1.t (i int);
CREATE TABLE
postgres=# insert into s1.t values (1);
INSERT 0 1

# resolves to table 't'
postgres=# select * from t;
 i 
---
 1
(1 row)

postgres=# create temp view t as select 2 as i;
CREATE VIEW

# resolves to temp view 't'
postgres=# select * from t;
 i 
---
 2
(1 row)

# resolves to temp view 't'
postgres=# insert into t values (1);
2019-12-05 21:40:47.229 EST [5451] ERROR:  cannot insert into view "t"
2019-12-05 21:40:47.229 EST [5451] DETAIL:  Views that do not select from a single table or view are not automatically updatable.
```